prompt_toolkit/shortcuts/progress_bar/base.py

```diff
@@ -370,8 +382,15 @@ def done(self, value: bool) -> None:
         # If done then store the stop_time, otherwise clear.
         self.stop_time = datetime.datetime.now() if value else None
 
-        if value and self.remove_when_done:
-            self.progress_bar.counters.remove(self)
+        if value:
+            if callable(self.remove_when_done):
+                # Register a background task to run the remove_when_done callable.
```

Comment:
This can be written as:
```python
self.progress_bar._app_loop.call_soon_threadsafe(
    lambda: self.progress_bar.app.create_background_task(self._remove_when_done_async()))
````

I prefer not to pass a loop argument to ``create_background_task``, because it's not thread safe and should not be called from a different thread to begin with. The only reason we have that function is to spawn coroutines in the current loop, and make sure they are cancelled when the application terminates.

Reply:
ah ok will do